File: food-delivery-app/lib/features/authentication/models/deliverer/residency_info.dart
import 'package:food_delivery_app/data/services/reflect.dart';
import 'package:food_delivery_app/utils/helpers/helper_functions.dart';

@reflector
@jsonSerializable
class DelivererResidencyInfo {
  String? deliverer;
  final bool? isSameAsCI;
  final String? city;
  final String? district;
  final String? ward;
  final String? address;
  final String? taxCode;
  final String? email;

  DelivererResidencyInfo({
    this.deliverer,
    this.isSameAsCI,
    this.city,
    this.district,
    this.ward,
    this.address,
    this.taxCode,
    this.email,
  });

  DelivererResidencyInfo.fromJson(Map<String, dynamic> json)
      : deliverer = json['deliverer'],
        isSameAsCI = json['is_same_as_ci'],
        city = json['city'],
        district = json['district'],
        ward = json['ward'],
        address = json['address'],
        taxCode = json['tax_code'],
        email = json['email'];

  Map<String, dynamic> toJson({bool patch = false}) {
    final data = {
      'deliverer': deliverer,
      'is_same_as_ci': isSameAsCI,
      'city': city,
      'district': district,
      'ward': ward,
      'address': address,
      'tax_code': taxCode,
      'email': email,
    };

    if (patch) {
      data.removeWhere((key, value) => value == null);
    }

    return data;
  }


  @override
  String toString() {
    return THelperFunction.formatToString(this);
  }
}

File: food_delivery_backend/deliverer/serializers/residency_info.py
from rest_framework import serializers
from deliverer.models import ResidencyInfo

class BaseResidencyInfoSerializer(serializers.ModelSerializer):
    class Meta:
        model = ResidencyInfo
        fields = [
            'id',
            'deliverer',
            'is_same_as_ci',
            'city',
            'district',
            'ward',
            'address',
            'tax_code',
            'email',
        ]
        read_only_fields = ['id']

class ResidencyInfoSerializer(BaseResidencyInfoSerializer):
    pass  

class UpdateResidencyInfoSerializer(BaseResidencyInfoSerializer):
    class Meta(BaseResidencyInfoSerializer.Meta):
        read_only_fields = ['id', 'deliverer']  
